### DISCUSSION TITLE
쿠폰 이벤트 참여 API 부하테스트

### DISCUSSION THREAD

**Discussion by kmularise** [dailyge/dailyge-server#190] [Decision]
## 📝 연관 PR
- [[DLG-252] 쿠폰 발급 요청 묶음을 처리하는 API를 개발한다.](https://github.com/dailyge/dailyge-server/pull/140)
- [[DLG-313] 쿠폰 이벤트 참여 신청 시, 이벤트 도메인 로직을 추가한다.](https://github.com/dailyge/dailyge-server/pull/169)

<br/><br/>

## ❓ 작성 이유

쿠폰 이벤트 참여 API에 관한 부하테스트 진행 시 마주했었던 문제 상황과 부하테스트 결과에 대해 회의한 내용을 기록하고 공유하고자 작성했습니다.

<br/><br/>

## 부하테스트 1차 결과
![1500](https://github.com/user-attachments/assets/6ef37f2a-0bcf-4bec-8b74-e6d1b71c69ce)

가상 사용자 1500일때, TPS는 4200대가 나왔지만, 응답 성공률이 98.0%로 기존에 설정했던 서버 가용성 기준 99.9%에 비해 낮게 나왔음을 알게 되었고, 실패한 요청에서 Socket Exception이 발생했음을 알게 되었습니다. 이러한 Socket Exception이 발생한 원인을 회의한 결과, 테스트를 실행하는 환경에서 네트워크 문제가 발생할 가능성이 있음을 알게 되었습니다. 이에 따라 다른 환경에서 부하테스트를 해보기로 결정했습니다. 

<br/><br/>

## 부하테스트 2차 결과

네트워크가 안정적인 환경에서 테스트를 실행하는 방식으로 바뀌고 가상 사용자가 각각 1000명, 1200명일 때 5분동안 테스트해보았습니다. 쿠폰 선착순 이벤트는 1분에서 2분 안에 종료된다고 생각하여 5분동안 부하를 넣는 것이 충분하다고 생각했기 때문입니다.

### 가상 사용자 1000명
<img width="793" alt="image" src="https://github.com/user-attachments/assets/32c60d26-aaec-4ee0-a132-5c9dba073dfa">

가상 사용자 1000명일 때 평균 TPS는 2,454.8입니다. 이때 응답 성공률은 99.96%로 가용성 기준 99.9%를 만족시킴을 알 수 있었습니다. 네트워크가 안정적인 환경에서 실행하자 응답 성공률이 훨씬 높아짐을 볼 수 있었습니다. 이렇게 네트워크 안정성이 부하테스트 성능 차이를 보이는 이유는 지연 시간(Latency)와 대역폭(bandwidth)에 영향을 주기 때문이라고 생각합니다.

<img width="793" alt="image" src="https://github.com/user-attachments/assets/ad4404b1-309a-492e-b03d-ddbf78b83fc6">

평균 응답속도는 406.61ms입니다. 사용자가 250ms부터 지연을 느끼기 시작하므로 성능 개선을 하게 되면 사용자 경험을 향상시킬 것이라 생각합니다.

<img width="793" alt="image" src="https://github.com/user-attachments/assets/1a66b4e2-0e3d-4985-a018-091f1e75e152">

응답 실패는 테스트 초반에는 거의 오류가 나타나지 않다가 간헐적으로 일부 군간에서 연속적으로 오류가 발생하는 패턴을 보입니다. 회의한 결과, Redis의 메모리 사용률과 연결풀에서 이상이 없는 것으로 보아, GC 발생으로 인해 애플리케이션이 멈추는 현상이 원인 중 하나라는 결론을 도출했습니다. 이를 위해 추후에 GC 튜닝을 거쳐할 것 같습니다.

<br/><br/>

### 가상 사용자 1200명
<img width="793" alt="image" src="https://github.com/user-attachments/assets/7b7164cc-cb44-4979-8fae-a760d3bb3b35">
<img width="793" alt="image" src="https://github.com/user-attachments/assets/5ade8e28-779e-4dfa-a53d-8cb944e067b8">

<br/><br/>

## 스케줄러 제어와 RabbitMQ의 도입 필요성
부하테스트를 위해 쿠폰 이벤트들을 Redis로 이동시키는 작업을 동적으로 스케줄러에 등록해야 했습니다. 이때 저희가 접근할 수 있는 주소는 로드밸런서 주소여서 애플리케이션 서버의 제어가 힘들었습니다. 이를 해결하기 위해 회의한 결과, RabbitMQ를 추후에 도입하자는 결론이 나왔습니다. RabbitMQ를 이용하면 메시지 큐에 작업을 등록하고, 각 애플리케이션 서버가 필요한 작업을 가져가서 처리할 수 있습니다.

**Comment by devjun10** [comment]
잘 봤습니다 ㅎㅎ 테스트 하면서 너무 재밌었네요. 수퍼노바님 💃 공유기나 네트워크 환경이 너무 불안정해서 다음엔 조금 더 나은 곳에서.. 언젠간...

**Comment by devjun10** [comment]
이거 다시 생각해봐도 네트워크만 좀 멀쩡하면 TPS 훨씬 잘 나올 것 같은데 뭔가 아쉽네요?

**Reply by kmularise** [reply]
네트워크 따라 달라져서 아쉽네요! 2차 부하테스트때 한번 생각해봐요!
